File: 프로그래머스/4/72416. 매출 하락 최소화/매출 하락 최소화.js
function solution(sales, links) {
    const team = new Map();
    for (const [head, member] of links) {
        if (!team.has(head)) team.set(head, []);
        team.get(head).push(member);
    }
    // dp[i][0]: i가 불참할 때, i의 하위 팀들의 최소 비용
    // dp[i][1]: i가 참석할 때, i의 하위 팀들의 최소 비용
    const dp = Array.from({ length: sales.length + 1 }, () => [0, 0]);
    
    function dfs(curr) {
        dp[curr][0] = 0;
        dp[curr][1] = sales[curr - 1];

        if (!team.has(curr)) return; // 팀장이 아닌 경우
        
        let totalSum = 0;
        let isAnyMemberAttending = false;
        let minDiff = Infinity;
        
        for (const member of team.get(curr)) {
            dfs(member); // 자식 노드 먼저 계산

            if (dp[member][0] < dp[member][1]) {
                totalSum += dp[member][0];
                // 나중에 강제로 한 명 보낼 때를 대비해 비용 차이 저장
                minDiff = Math.min(minDiff, dp[member][1] - dp[member][0]);
            } else {
                totalSum += dp[member][1];
                isAnyMemberAttending = true;
            }
        }
        // 팀장이 참석하면 팀원들은 가든 말든 상관없음
        dp[curr][1] += totalSum;
        // 팀장이 참석 안하면 -> 자기가 가는게 이득인 팀원이 있으면 팀원의 비용 / 없으면 강제로 하나 보냄.
        dp[curr][0] = isAnyMemberAttending ? totalSum : totalSum + minDiff
    }
    
    dfs(1);
    return Math.min(dp[1][0], dp[1][1]);
}

File: 프로그래머스/4/72416. 매출 하락 최소화/매출 하락 최소화.py
import sys

# 재귀 깊이 제한 해제
sys.setrecursionlimit(300000)

def solution(sales, links):
    team = {}
    for parent, child in links:
        if parent not in team:
            team[parent] = []
        team[parent].append(child)
    
    # dp[노드][0]: 불참, dp[노드][1]: 참석
    n = len(sales)
    dp = [[0, 0] for _ in range(n + 1)]
    
    def dfs(curr):
        dp[curr][0] = 0
        dp[curr][1] = sales[curr - 1]
        
        # 자식(팀원)이 없는 리프 노드인 경우 종료
        if curr not in team:
            return
        
        total_sum = 0
        is_any_member_attending = False
        min_diff = float('inf')
        
        for member in team[curr]:
            dfs(member) # 자식부터 계산 (Bottom-up)
            
            # 자식 노드에서 [안 갈 때, 갈 때] 중 더 저렴한 것 선택
            if dp[member][0] < dp[member][1]:
                total_sum += dp[member][0]
                # 강제 동원을 대비한 최소 비용 차이 저장
                min_diff = min(min_diff, dp[member][1] - dp[member][0])
            else:
                total_sum += dp[member][1]
                is_any_member_attending = True
                
        # 팀장이 참석하면 팀원들은 최적의 선택(가든 말든)을 합산한 total_sum만 더해주면 됨
        dp[curr][1] += total_sum
        # 팀장이 참석 안 할 때: 한 명이라도 이미 참석 중이면 그대로, 아니면 min_diff를 더함
        dp[curr][0] = total_sum if is_any_member_attending else total_sum + min_diff

    dfs(1)
    return min(dp[1][0], dp[1][1])
